### PR TITLE
tp-qemu: Add a framework for whql related cases when driver in use.

### PR DESCRIPTION
**Pull request by CongSmile**
Signed-off-by: Cong Li <coli@redhat.com>

### CODE REVIEW

File: qemu/tests/driver_in_use.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                env[bg_stress_run_flag] = False
+                stress_thread = utils.InterruptedThread(
+                    utils_test.run_virt_sub_test, (test, params, env),
+                    {"sub_type": bg_stress_test})
+                stress_thread.start()
+                if stress_thread.is_alive():
+                    try:
+                        stop_time = time.time() + 600
+                        while time.time() < stop_time:
+                            status = check_bg_status(session, target)
```
Comment (xutian) [inline qemu/tests/driver_in_use.py]: s/status/running/

File: qemu/tests/driver_in_use.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+                            if status:
+                                env["bg_status"] = 1
+                                break
+                    except Exception:
+                        raise exceptions.TestFail("Backgroud test %s is not"
+                                                  "alive!" % bg_stress_test)
+                if not utils_misc.wait_for(lambda: env.get(bg_stress_run_flag),
+                                           wait_time, 0, 5,
+                                           "Wait %s test start" % bg_stress_test):
+                    err = "Fail to start netperf test between guest and host"
```
Comment (suqinhuang) [inline qemu/tests/driver_in_use.py]: Hi coli,

Why netperf test here?

File: qemu/tests/driver_in_use.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                target = params.get("target", "")
+                bg_stress_run_flag = params.get("bg_stress_run_flag")
+                env[bg_stress_run_flag] = False
+                stress_thread = utils.InterruptedThread(
+                    utils_test.run_virt_sub_test, (test, params, env),
+                    {"sub_type": bg_stress_test})
+                stress_thread.start()
+                if stress_thread.is_alive():
+                    try:
+                        stop_time = time.time() + 600
```
Comment (suqinhuang) [inline qemu/tests/driver_in_use.py]: can we set 600 configuable

File: qemu/tests/driver_in_use.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    3) run subtest before / during / after background stress test.
+    4) clear the driver verifier.
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    def check_bg_status(session, target):
+        """
```
Comment (suqinhuang) [inline qemu/tests/driver_in_use.py]: Hi coli,
seems you are checking driver here, not background test

File: qemu/tests/driver_in_use.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        bg_stress_test = params.get("run_bgstress")
+        if bg_stress_test:
+            try:
+                error_context.context("Run test %s background" % bg_stress_test,
+                                      logging.info)
+                stress_thread = ""
+                wait_time = float(params.get("wait_bg_time", 60))
+                target = params.get("target", "")
+                bg_stress_run_flag = params.get("bg_stress_run_flag")
```
Comment (suqinhuang) [inline qemu/tests/driver_in_use.py]: Hi coli,
Why bg_stress_run_flag, seems it's always False in your design

File: qemu/tests/driver_in_use.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    4) clear the driver verifier.
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    def check_bg_status(session, target):
+        """
+        Check the backgroud test status in guest.
```
Comment (suqinhuang) [inline qemu/tests/driver_in_use.py]: Hi coli,
seems you are checking driver here, not background test

Comment (CongSmile) [inline qemu/tests/driver_in_use.py]: It checks the stress status in guest here.

the target is the process running in the guest.

**Comment by CongSmile** [comment]
Based on #447

File: qemu/tests/driver_in_use.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+                    utils_test.run_virt_sub_test, (test, params, env),
+                    {"sub_type": bg_stress_test})
+                stress_thread.start()
+                if stress_thread.is_alive():
+                    try:
+                        stop_time = time.time() + wait_time
+                        while time.time() < stop_time:
+                            status = check_bg_running(session, target)
+                            if status:
+                                env["bg_status"] = 1
```
Comment (suqinhuang) [inline qemu/tests/driver_in_use.py]: Hi,
Where is bg_status is used?

File: qemu/tests/driver_in_use.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+                                break
+                    except Exception:
+                        raise exceptions.TestFail("Backgroud test %s is not"
+                                                  "alive!" % bg_stress_test)
+                if not utils_misc.wait_for(lambda: env.get(bg_stress_run_flag),
+                                           wait_time, 0, 5,
+                                           "Wait %s test start" % bg_stress_test):
+                    err = ("Fail to start %s test between guest and host" %
```
Comment (suqinhuang) [inline qemu/tests/driver_in_use.py]: hi,
why "test between guest and host" ?

File: qemu/tests/driver_in_use.py (hunk 5 above)
Comment (CongSmile) [inline qemu/tests/driver_in_use.py]: Hi shuang,

+                stress_thread = utils.InterruptedThread(
+                    utils_test.run_virt_sub_test, (test, params, env),
+                    {"sub_type": bg_stress_test})

In the background case, the flag will be set.

Refer (from netperf_stress):
            # here when set a run flag, when other case call this case as a
            # subprocess backgroundly, can set this run flag to False to stop
            # the stress test.
            env["netperf_run"] = True

**Comment by CongSmile** [comment]
@xutian @vipmike007 

I have updated the code for balloon_stress test.
Could you help review the case again?

I will update the code according to your comments.

Thanks.

File: qemu/tests/driver_in_use.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+                        raise exceptions.TestFail("Backgroud test %s is not"
+                                                  "alive!" % bg_stress_test)
+                if not utils_misc.wait_for(lambda: env.get(bg_stress_run_flag),
+                                           wait_time, 0, 5,
+                                           "Wait %s test start" % bg_stress_test):
+                    err = "Fail to start %s test" % bg_stress_test
+                    raise exceptions.TestError(err)
+                return stress_thread
+            except:
+                pass
```
Comment (xutian) [inline qemu/tests/driver_in_use.py]: why you ignore all exception here ?

**Comment by CongSmile** [comment]
Hi @xutian 

I have updated the code according to your comments, could you help review it again?

Thanks.

**Comment by suqinhuang** [comment]
LGTM

Acked-by: Suqin Huang

File: qemu/tests/driver_in_use.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+                        raise exceptions.TestFail("Backgroud test %s is not"
+                                                  "alive!" % bg_stress_test)
+                if not utils_misc.wait_for(lambda: env.get(bg_stress_run_flag),
+                                           wait_time, 0, 5,
+                                           "Wait %s test start" % bg_stress_test):
+                    err = "Fail to start %s test" % bg_stress_test
+                    raise exceptions.TestError(err)
+                return stress_thread
+            except Exception:
+                raise exceptions.TestFail(err)
```
Comment (xutian) [inline qemu/tests/driver_in_use.py]: what will be include in your err?

Comment (CongSmile) [inline qemu/tests/driver_in_use.py]: Hi @xutian 

I have updated the code like the following code:
+            except Exception:
+                raise exceptions.TestFail("Background case %s start failed!" %
+                                          bg_stress_test)

I want to give some info to specify the background case failed.

What do you think?

Thanks.

File: qemu/tests/driver_in_use.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        sub_type = params.get("sub_test")
+        if sub_type:
+            try:
+                msg = "Run sub test %s %s" % (sub_type, run_bg_flag)
+                error_context.context(msg, logging.info)
+                utils_test.run_virt_sub_test(test, params, env, sub_type)
+            except Exception:
+                raise exceptions.TestFail("Test failed when run %s!" % sub_type)
+
+    driver = params.get("driver_name")
```
Comment (xutian) [inline qemu/tests/driver_in_use.py]: it looks driver_name is must, so I think params["driver_name"] is necessary here.

File: qemu/tests/driver_in_use.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        run_bg_flag = params.get("run_bg_flag")
+        bg_stress = params.get("run_bgstress")
+        wait_time = float(params.get("wait_bg_time", 60))
+        suppress_exception = params.get("suppress_exception", "False")
+        if run_bg_flag == "before_bg_test":
+            run_subtest(run_bg_flag)
+            stress_thread = bg_stress_test()
+            if stress_thread:
+                stress_thread.join(timeout=timeout,
+                                   suppress_exception=suppress_exception)
```
Comment (xutian) [inline qemu/tests/driver_in_use.py]: suppress_exception is string not a bool value, so here will be always True.

File: qemu/tests/driver_in_use.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+        """
+        Run backgroud test.
+
+        :return: return the background case thread if it's successful;
+                 else raise error.
+        """
+        bg_stress_test = params.get("run_bgstress")
+        if bg_stress_test:
+            try:
+                error_context.context("Run test %s background" % bg_stress_test,
```
Comment (xutian) [inline qemu/tests/driver_in_use.py]: move out try section.

File: qemu/tests/driver_in_use.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+                                      logging.info)
+                stress_thread = ""
```
Comment (xutian) [inline qemu/tests/driver_in_use.py]: None looks better that empty string

File: qemu/tests/driver_in_use.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+        :param session: VM session.
+        :param target_process: Background process running in guest.
+        :return: return True if find the driver name;
+                 else return False
+        """
+        list_cmd = "wmic process where name='%s' list" % target_process
+        try:
+            output = session.cmd_output_safe(list_cmd, timeout=120)
+            check_reg = re.compile(r"%s" % target_process, re.I | re.M)
+            if bool(check_reg.findall(output)):
```
Comment (xutian) [inline qemu/tests/driver_in_use.py]: no need to convert here

File: qemu/tests/driver_in_use.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+                return True
+        except Exception, err:
+            raise exceptions.TestFail("Check process error: %s" % err)
```
Comment (xutian) [inline qemu/tests/driver_in_use.py]: unnecessary  try .. except here.

File: qemu/tests/driver_in_use.py (hunk 12 above)
Comment (CongSmile) [inline qemu/tests/driver_in_use.py]: Done.

**Comment by CongSmile** [comment]
Hi @xutian 

I have update the code according to your comments, could you help review again?

Thanks.

File: qemu/tests/driver_in_use.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+        elif run_bg_flag == "during_bg_test":
+            stress_thread = bg_stress_test()
+            stop_time = time.time() + wait_time
+            while time.time() < stop_time:
+                if (env["bg_status"] == 1):
+                    # When case "balloon_stress" as a background case, need
+                    # make sure that:
+                    # 1. balloon is doing in qemu side.
+                    # 2. video is palying in guest side.
+                    if bg_stress == "balloon_stress" and (env["balloon_test"] == 1):
```
Comment (xutian) [inline qemu/tests/driver_in_use.py]: useless ()

File: qemu/tests/driver_in_use.py [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+            stress_thread = None
+            wait_time = float(params.get("wait_bg_time", 60))
+            target_process = params.get("target_process", "")
+            bg_stress_run_flag = params.get("bg_stress_run_flag")
+            env[bg_stress_run_flag] = False
+            stress_thread = utils.InterruptedThread(
+                utils_test.run_virt_sub_test, (test, params, env),
+                {"sub_type": bg_stress_test})
+            stress_thread.start()
+            if stress_thread.is_alive():
```
Comment (xutian) [inline qemu/tests/driver_in_use.py]: if utils_misc.wait_for(stress_thread.is_alive,  10); then check_bg_running(session, target_process) else: raise error.  logical will looks more clearly. right ?

**Comment by CongSmile** [comment]
Hi @xutian 

I have updated the code, could you help review it again?

Thanks.

**Comment by xutian** [comment]
works for me, merged.  thanks @CongSmile and @suqinhuang

**Pull request merged**